include/unifex/await_transform.hpp

```diff
@@ -0,0 +1,186 @@
+
+#if UNIFEX_NO_COROUTINES
+# error "Coroutine support is required to use <unifex/coroutine_concepts.hpp>"
+#endif
+
+#include <unifex/async_trace.hpp>
+#include <unifex/coroutine_concepts.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/type_traits.hpp>
+#include <unifex/manual_lifetime.hpp>
+
+#include <cassert>
+#include <exception>
+#include <optional>
+#include <type_traits>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _await_transform {
+
+template <typename Promise, typename Sender>
+class _as_awaitable {
+  using value_t = single_value_result_t<Sender>;
+  enum class state { empty, value, exception, done };
+
+  class receiver {
+  public:
+    explicit receiver(_as_awaitable* op, coro::coroutine_handle<Promise> continuation) noexcept
+    : op_(op)
+    , continuation_(continuation)
+    {}
+
+    receiver(receiver&& r) noexcept
+    : op_(std::exchange(r.op_, nullptr))
+    , continuation_(std::exchange(r.continuation_, nullptr))
+    {}
+
+    template(class... Us)
+      (requires constructible_from<value_t, Us...>)
+    void set_value(Us&&... us) &&
+        noexcept(std::is_nothrow_constructible_v<value_t, Us...>) {
+      unifex::activate_union_member(op_->value_, (Us&&) us...);
+      op_->state_ = state::value;
+      continuation_.resume();
+    }
+
+    void set_error(std::exception_ptr eptr) && noexcept {
+      unifex::activate_union_member(op_->exception_, std::move(eptr));
+      op_->state_ = state::exception;
+      continuation_.resume();
+    }
+    
+    void set_done() && noexcept {
+      op_->state_ = state::done;
+      continuation_.resume();
+    }
+
+    template(typename CPO)
+      (requires is_receiver_query_cpo_v<CPO> AND is_callable_v<CPO, const Promise&>)
+    friend auto tag_invoke(CPO cpo, const receiver& r)
+        noexcept(is_nothrow_callable_v<CPO, const Promise&>)
+        -> callable_result_t<CPO, const Promise&> {
+      const Promise& p = r.continuation_.promise();
+      return std::move(cpo)(p);
+    }
+
+    template <typename Func>
+    friend void
+    tag_invoke(tag_t<visit_continuations>, const receiver& r, Func&& func) {
+      visit_continuations(r.continuation_.promise(), (Func&&)func);
+    }
+
+  private:
+    _as_awaitable* op_;
+    coro::coroutine_handle<Promise> continuation_;
+  };
+
+public:
+  explicit _as_awaitable(Sender&& sender, coro::coroutine_handle<Promise> h)
+    noexcept(is_nothrow_connectable_v<Sender, receiver>)
+  : op_(unifex::connect((Sender&&)sender, receiver{this, h}))
+  {}
+
+  ~_as_awaitable() {
+    switch(state_) {
+    case state::value:
+      unifex::deactivate_union_member(value_);
+      break;
+    case state::exception:
+      unifex::deactivate_union_member(exception_);
+      break;
+    default:;
+    }
+  }
+
+  bool await_ready() const noexcept {
+    return false;
+  }
+
+  void await_suspend([[maybe_unused]] coro::coroutine_handle<Promise> h) noexcept {
```

Comment:
Typically you don't need `[[maybe_unused]]` if you don't name the argument.
```suggestion
  void await_suspend(coro::coroutine_handle<Promise>) noexcept {
```